projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-list-users-endpoint.php

```diff
@@ -171,87 +171,107 @@ public function callback( $path = '', $blog_id = 0 ) {
+		$is_multisite    = is_multisite();
+		$user_query      = new WP_User_Query( $query );
+
+		// Get users.
+		foreach ( $user_query->get_results() as $u ) {
+			$the_user = $this->get_author( $u, true );
+			if ( $the_user && ! is_wp_error( $the_user ) ) {
+				$userdata        = get_userdata( $u );
+				$the_user->roles = ! is_wp_error( $userdata ) ? array_values( $userdata->roles ) : array();
+				if ( $is_multisite ) {
+					$the_user->is_super_admin = user_can( $the_user->ID, 'manage_network' );
 				}
+				$users[] = $the_user;
+			}
+		}
+
+		// Get viewers.
+		if ( $include_viewers ) {
+			$page    = ( (int) ( $args['offset'] / $args['number'] ) ) + 1;
+			$viewers = get_private_blog_users(
+				$blog_id,
+				array(
+					'page'     => $page,
+					'per_page' => $args['number'],
+				)
 			);
+
+			// Returns author object. See `WPCOM_JSON_API_Endpoint::get_author` in `class.json-api-endpoints.php`.
+			$viewers = array_map( array( $this, 'get_author' ), $viewers );
+
+			// Search by username or email.
+			if ( ! empty( $args['search'] ) ) {
+				$viewers = array_filter(
+					$viewers,
+					function ( $viewer ) use ( $args ) {
+						// Convert to WP_User so expected fields are available.
+						$wp_viewer = new WP_User( $viewer->ID );
+						// remove special database search characters from search term
+						$search_term = str_replace( '*', '', $args['search'] );
+						return ( str_contains( $wp_viewer->user_login, $search_term ) || str_contains( $wp_viewer->user_email, $search_term ) || str_contains( $wp_viewer->display_name, $search_term ) );
+					}
+				);
+			}
+
+			$viewer_ids     = array();
+			$unique_viewers = array();
+			// Create a lookup array of viewer IDs.
+			foreach ( $viewers as $viewer ) {
+				$viewer_ids[ $viewer->ID ] = true;
+			}
+
+			// Add viewer role to users who are also viewers.
+			foreach ( $users as $user ) {
+				if ( isset( $viewer_ids[ $user->ID ] ) && ! in_array( 'viewer', $user->roles, true ) ) {
+					$user->roles[] = 'viewer';
+					// Mark this user so we don't add them again.
+					$viewer_ids[ $user->ID ] = false;
+				}
+			}
+
+			// Add viewer role to viewers and remove duplicates.
+			foreach ( $viewers as $viewer ) {
+				if ( isset( $viewer_ids[ $viewer->ID ] ) && true === $viewer_ids[ $viewer->ID ] ) {
+					$viewer->roles[]  = 'viewer';
+					$unique_viewers[] = $viewer;
+				}
+			}
+
+			// Reassign the viewers array to the unique viewers array.
+			$viewers = $unique_viewers;
 		}
 
-		$return = array();
 		foreach ( array_keys( $this->response_format ) as $key ) {
 			switch ( $key ) {
 				case 'found':
 					$user_count = (int) $user_query->get_total();
-
-					$viewer_count = 0;
 					if ( $include_viewers ) {
-						if ( empty( $args['search'] ) ) {
-							$viewer_count = (int) get_count_private_blog_users( $blog_id );
-						} else {
-							$viewer_count = count( $viewers );
-						}
+						$viewer_count     = count( $viewers );
```

Comment:
This won't be accurate due to legacy duplication. See: https://github.com/Automattic/wp-calypso/pull/100674#issuecomment-2785687150